web/server/codechecker_server/api/mass_store_run.py

```diff
@@ -191,6 +229,14 @@ def get_blame_file_data(
     return blame_info, remote_url, tracking_branch
 
 
+def checker_name_for_report(report: Report) -> Tuple[str, str]:
+    analyzer_name, checker_name = \
+        getattr(report, "analyzer_name"), getattr(report, "checker_name")
+    analyzer_name = analyzer_name if analyzer_name else "UNKNOWN"
+    checker_name = checker_name if checker_name else "NOT FOUND"
+    return (analyzer_name, checker_name)
```

Comment:
Unfortunately, the new version is **not** equivalent to what is in the patch. (Note that type **hints** are just hints, there is no guarantee that indeed a conforming `Report` object is passed to this function!) I don't have clear memories of this anymore due to having worked multiple months on this feature, but I would wager that the [mentioned raw Cppcheck plist test file](https://github.com/Ericsson/codechecker/blob/v6.23.0/web/tests/functional/cppcheck/test_proj/divide_zero.plist) was the way to trigger the inequivalence.

The coalescing `x or y` will only return `y` if `x` is successfully evaluated, and returns falsy. A direct `.` operator, however, may result in an `AttributeError`, something that is only coalesced by `getattr()`'s implicit `default = None` parameter.

```py3
$ python3
Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class C:
...   def __init__(self):
...     self.member = "X"
...     self.member2 = None
...
>>> c = C()
>>> (c.member or "default 1", c.member2 or "default 2")
('X', 'default 2')
>>>
>>>
>>>
>>> a = object()
>>> (a.member or "default 1", a.member2 or "default 2")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'object' object has no attribute 'member'
```

Reply:
I would say that we should rely on type-hints, that's why we use them. If we expect a `Report` object, then we may trust it has the required fields. This is the exact situation what I meant at a previous comment: if we are distrustful with the caller, it complicates a lot at the callee side. So, if this function gets something else than a `Report`, then we should fix that call. Another option is to extend the type-hint to indicate other possible parameter types too, but I would prefer the first option.

Reply:
> we should fix that call

It's a call through the API, which involves what is being put into the mass-store ZIP. **If** the bogus behaviour here (fields being unknown) happen because of plists NOT coming from the report-converter are being uploaded, then this means that a full-scale schema validation logic must be developed into the **`store`** (and the _`parse`_?) command such that it doesn't allow upgrading.

The exact details of what format of `plist`s is accepted over the API is not documented anywhere, so we might be breaking a lot of downstream projects (VSCode Plugin? Eclipse Plugin? Third-party users who create their own `plist`s from analyzers not passed through the report converter?) as well.

Or should we just let the `AttributeError` escape here and explode the entire storage of the user, and let the user figure out why their reports are broken?

Reply:
Nevertheless, I checked all the edge cases I could find and figured out a way to simplify the related logic in multiple places.